src/openrct2-ui/windows/Footpath.cpp

```diff
@@ -69,30 +70,10 @@ namespace OpenRCT2::Ui::Windows
         ObjectEntryIndex type;
         CoordsXYZ position;
         uint8_t slope;
-        uint8_t flags;
+        FlagHolder<uint8_t, ProvisionalPathFlag> flags;
```

Comment:
Hmm, I think this is overly verbose, to be honest. In terms of type, I'd vastly prefer something a solid type like `ProvisionalPathFlags` for the container. To the caller, the underlying types shouldn't need to be specified, imo, which is necessitated by this kind of templating.

We've talked about adopting [OpenLoco's EnumFlags.hpp](https://github.com/OpenLoco/OpenLoco/blob/master/src/Core/include/OpenLoco/Core/EnumFlags.hpp) to this end before. I still think that would be the better solution, including the usual bit operators. We could add more methods if preferred.

Reply:
I really dislike those, because those require you to specify the type as an enum, while not working as one. It is a _holder_ for flags - two distinct things - and OpenLoco just smashes them together in a way that doesn’t really make sense IMO.

Reply:
> I really dislike those OpenLoco “enums”, because those require you to specify the type as an enum, while not working as one. It is a _holder_ for flags - so two things - and OpenLoco just smashes them together in a way that doesn’t really make sense IMO.

Hmm, I don't see it that way at all. If anything, it becomes more like a `class` with `enum` properties. This is just much like how methods on enums work in other languages, such as Rust and Java. But perhaps this is a discussion for the team chat.

> Also, I don’t see how this is more verbose - OpenLoco requires you to put `OPENLOCO_ENABLE_ENUM_OPERATORS` underneath the enum, which is no shorter.

Suppose you pass the flags to another function. Your example would necessitate spelling out the template parameters as well, i.e. `FlagHolder<uint8_t, ProvisionalPathFlag> flags`. The OpenLoco enums would just need the main class type, e.g. `ProvisionalPathFlags flags`.

Reply:
> I really dislike those OpenLoco “enums”, because those require you to specify the type as an enum, while not working as one. It is a _holder_ for flags - so two things - and OpenLoco just smashes them together in a way that doesn’t really make sense IMO.
> 
> Also, I don’t see how this is more verbose - OpenLoco requires you to put `OPENLOCO_ENABLE_ENUM_OPERATORS` underneath the enum, which is no shorter.

Your type also uses an enum to specify the bit index, so if the underlying type is the enum or an integer and shifted by an enum is to me no different at all. It is unfortunate that C++ has no builtin support to have type safe flags however a lot of the solutions actually tend to just use enum class like in OpenLoco, another reason to why I prefer the approach is that in Visual Studio actually displays the value properly in the debugger while your class eliminates that.

Reply:
One other reason I don’t like OpenLoco’s approach is that OpenLoco’s "enums" have built-in bitshifts, which makes cases like `RideMode` impossible - it is used to save the ride mode on a ride, but it is also used on the RTD for the collection of possible ride modes. There are probably more and better examples, but this was the first to come to mind.

Seeing as I do most refactors, I do think that the solution we pick should be one that I can work with. I have no issue taking on the job to change all of the existing occurrences, but I just can’t with OpenLoco’s approach. (And doing nothing is not a good idea either, what we currently have is just messy.)

Reply:
I understand the problem however I think that in your example we are actually dealing with two separate types. Given the problem I think the new type is acceptable, for the time being, however we should create an alias for those flags, something like `using ProvisionalPathFlag = FlagHolder<uint8_t, ProvisionalPathFlag>;` is probably cleaner.

Reply:
I can work with that. :+1:
I pushed an update.